util/hot_backup_test.cc

```diff
@@ -0,0 +1,600 @@
+
+    return(ret_val);
+}   // main
+
+
+namespace leveldb {
+
+
+/**
+ * Wrapper class for tests.  Holds working variables
+ * and helper functions.
+ */
+class HotBackupTester : public HotBackup
+{
+public:
+    std::string m_DBName;
+    std::string m_Trigger;
+
+    HotBackupTester()
+    {
+        m_DBName = test::TmpDir() + "/hot_backup";
+        m_Trigger = test::TmpDir() + "/trigger";
+        gHotBackup=this;
+    };
+
+    virtual ~HotBackupTester()
+    {
+        ResetHotBackupObject();
+    };
+
+    virtual const char * GetTriggerPath() {return(m_Trigger.c_str());};
+
+    void ClearAllBackups(
+        Options & InOptions)
+    {
+        int loop;
+        Options options;
+
+        // clear old if exists
+        for (loop=0; loop<=config::kNumBackups; ++loop)
+        {
+            options=InOptions;
+
+            SetBackupPaths(options, loop);
+            DestroyDB("", options);
+        }   // if
+
+        options.tiered_slow_level=4;
+        options.tiered_fast_prefix=m_DBName + "/fast";
+        options.tiered_slow_prefix=m_DBName + "/slow";
+        DestroyDB("", options);
+    }   // ClearAllBackups
+
+};  // class HotBackupTester
+
+
+/**
+ * Initial code used the existance of /etc/riak/hot_backup file
+ *  as flag to start a backup.
+ */
+TEST(HotBackupTester, FileTriggerTest)
+{
+    char * dup_path, *path;
+    int ret_val;
+    bool ret_flag;
+    uint64_t perf_before, perf_after;
+
+    perf_before=gPerfCounters->Value(ePerfSyslogWrite);
+
+    // cleanup anything existing, likely fails
+    ///  hmm, should there be a way to move this trigger
+    ///  to a "safe" area like /tmp?
+    unlink(GetTriggerPath());
+
+    // does parent path exist?
+    //  bypass test if it does not ... Travis CI and
+    //  other users might not be able to access /etc/riak
+    dup_path=strdup(GetTriggerPath());
+    path=dirname(dup_path);
+    ret_val=access(path, R_OK | W_OK);
+    ASSERT_TRUE(-1!=ret_val);
+
+    // is a trigger seen (hope not)
+    ret_flag=HotBackup::IsTriggerSet();
+    ASSERT_TRUE(!ret_flag);
+
+    // make a trigger
+    ret_val=open(GetTriggerPath(), O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
+    ASSERT_TRUE(-1!=ret_val);
+    close(ret_val);
+
+    // test the trigger
+    ret_flag=HotBackup::IsTriggerSet();
+    ASSERT_TRUE(ret_flag);
+
+    // pretend to be back process
+    /// schedule twice
+    HotBackup::HotBackupScheduled();
+    HotBackup::HotBackupScheduled();
+
+    // trigger still there?
+    ret_flag=HotBackup::IsTriggerSet();
+    ASSERT_TRUE(ret_flag);
+
+    // release one, trigger goes away
+    HotBackup::HotBackupFinished();
+    ret_flag=HotBackup::IsTriggerSet();
+    ASSERT_TRUE(!ret_flag);
+
+    // did our simulation create a syslog entry?
+    //  (bonus if you manually check /var/log/syslog for actual entries)
+    perf_after=gPerfCounters->Value(ePerfSyslogWrite) - perf_before;
+    ASSERT_TRUE( 1==perf_after );
+
+    // clean up second count.
+    HotBackupFinished();
```

Comment:
If I'm reading the code correctly (and I'll be the first to admit that I may not be doing so), it looks like this call is unnecessary. Here's what I _think_ happens: the first call to HotBackup::HotBackupFinished() (above) decrements m_JobsPending to 1 and calls ResetTrigger(), which deletes the trigger file and calls HotBackupFinished(). The second (recursive) call to HotBackupFinished() then decrements m_JobsPending to 0.

With all that said, I don't think this call to HotBackupFinished() causes a problem, but its associated comment is misleading (if my analysis above is correct).